lib/IPC.cpp

```diff
@@ -0,0 +1,266 @@
+/** @file IPC.cpp
+ *  @author Bao Vu (bao.vu@unikie.com)
+ *  @brief Include IPC class methods' definition & non-member functions' definition
+ */
+
+#include "lib/IPC.h"
+
+/******************************************************************************/
+/* IPC Class's methods ********************************************************/
+/******************************************************************************/
+
+unsigned long IPC::get_file_size() const
+{
+	return boost::filesystem::file_size(this->opts.file_name);
+}
+
+void IPC::cleanup()
+{
+	if (this->buffer != nullptr)
+	{
+		delete buffer;
+		std::cout << "Freed the buffer." << std::endl;
+	}
+	std::cout << "Closed file." << std::endl;
+	fs.close();
+}
+
+void IPC::print_members() const
+{
+	std::cout << "Mode: " << static_cast<std::underlying_type<IPCMode>::type>(opts.mode) << "\n"
+			  << "Protocol: " << static_cast<std::underlying_type<IPCProtocol>::type>(opts.protocol) << "\n"
+			  << "server_name: " << opts.server_name << "\n"
+			  << "file_name: " << opts.file_name << "\n"
+			  << "mem_size: " << opts.mem_size << "\n"
+			  << "timeout: " << timeout << "\n";
+	if (this->buffer == nullptr)
+		std::cout << "buffer: nullptr" << std::endl;
+	else
+		std::cout << "buffer: " << buffer << std::endl;
+}
+
+void IPC::auto_start()
+{
+	this->init();
+	if (this->opts.mode == IPCMode::SEND_MODE)
+		this->send();
+	if (this->opts.mode == IPCMode::RECEIVE_MODE)
+		this->receive();
+}
+
+void IPC::open_file()
+{
+	if (this->opts.mode == IPCMode::SEND_MODE)
+		this->fs.open(opts.file_name, std::fstream::in | std::fstream::binary);
+	if (this->opts.mode == IPCMode::RECEIVE_MODE)
+		this->fs.open(opts.file_name, std::fstream::out | std::fstream::trunc | std::fstream::binary);
+
+	if (this->fs.fail()) // Check logical error on I/O operation (close/open)
+		throw std::runtime_error("ERROR: " + this->opts.file_name + ": " + strerror(errno));
+}
+
+void IPC::write_to_file(char *data, std::streamsize &data_size)
+{
+	this->fs.write(data, data_size);
+	if (this->fs.bad()) // check read/writing error on i/o operation
+		throw std::runtime_error("ERROR: ostream::write().");
+}
+
+
+/******************************************************************************/
+/* Non-member function definitions ********************************************/
+/******************************************************************************/
+
+ipc_options_t *ipc_get_options(IPCMode mode, int argc, char *argv[])
+{
+	auto *options = new ipc_options_t;
```

Comment:
Is there a reason to use a pointer? 

If you use a pointer, you should delete it before exiting.

Reply:
Because it takes the point from main()

Reply:
So why you didn't use directly the object from main?

I mean I decided to use pointers, is there a reason?

Reply:
As you are using a new you are creating memory space for this.
When you are doing options->server_name = arg_list.at(++i); you are doing a copy. So I don't understand your point.

Reply:
I'm lost too ^^. My point is I like to keep my main() clean, everything should be handled in functions. And the pointer is what I get from main() then I use it directly and process it in my function.

Reply:
You can just use in your main:
ipc_options_t options = ipc_get_options(IPCMode::SEND_MODE, argc, argv);

And no pointer... just to avoid using new because sometimes you forgot to delete it (and that is the real issue)

Reply:
Okay. I got your point now ^^!